tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"bufio"
+	"encoding/binary"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Segment header fields constants.
+const (
+	headChunksFormatV1 = 1
+)
+
+// HeadReadWriter is for writing the Head block chunks to the disk
+// and access chunks via an mmapped file.
+type HeadReadWriter struct {
+	// Writer.
+	dirFile          *os.File
+	curFile          *os.File // Segment file being written to.
+	curFileStartTime time.Time
+	curFileSequence  int
+	wbuf             *bufio.Writer
+	wbufLock         sync.Mutex
+	n                int64 // Bytes written in current segment.
+	buf              [8]byte
+	segmentTime      time.Duration
+
+	// Reader
+	bs    map[int]ByteSlice // TODO: this should be a map of the file name and bytes slice as old slices will be deleted.
+	cs    map[int]io.Closer // Closers for resources behind the byte slices.
+	size  int64             // The total size of bytes in the reader.
+	bsMtx sync.RWMutex      // TODO: verify the usage of this
+	pool  chunkenc.Pool
+}
+
+const (
+	// DefaultHeadChunkSegmentTime is the default chunks segment time range.
+	DefaultHeadChunkSegmentTime = 15 * time.Minute
+	// HeadSegmentHeaderSize is the total size of the header for the segment file.
+	HeadSegmentHeaderSize = SegmentHeaderSize + 16
+	// HeaderMintOffset is the offset where the first byte of MinT for segment file exists.
+	HeaderMintOffset = SegmentHeaderSize
+	// HeaderMaxtOffset is the offset where the first byte of MaxT for segment file exists.
+	HeaderMaxtOffset = HeaderMintOffset + 8
+	// MaxSegmentSize is the max size of a segment file.
+	// Setting size to the max int32 as setting it to max int 64 crashes 64 systems too.
+	MaxSegmentSize = math.MaxInt32
+)
+
+// NewHeadReadWriter returns a new writer against the given directory
+// using the default segment time.
+func NewHeadReadWriter(dir string, pool chunkenc.Pool) (*HeadReadWriter, error) {
+	return newHeadReadWriter(dir, DefaultHeadChunkSegmentTime, pool)
+}
+
+func newHeadReadWriter(dir string, segmentTime time.Duration, pool chunkenc.Pool) (*HeadReadWriter, error) {
+	if segmentTime <= 0 {
+		segmentTime = DefaultHeadChunkSegmentTime
+	}
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, err
+	}
+	dirFile, err := fileutil.OpenDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	// mmap existing files
```

Comment:
Not a full sentence

Reply:
A cleanup which I missed, that comment isn't relevant now